Email candidates with unsuccessful applications before apply again

stevehook opened the pull request:
## Context

We want to send a call to action email to all candidates whose applications ended unsuccessfully before we released apply again.

## Changes proposed in this pull request

- [x] `CandidateMailer#apply_again_call_to_action` email
- [x] Mailer preview for the above
- [x] `SendApplyAgainCallToAction` service class to email all unsuccessful candidates

## Guidance to review

![image](https://user-images.githubusercontent.com/450843/83311483-4c7db900-a207-11ea-96e8-889304c6175f.png)

- I didn't want to over-engineer this as it's a one-off mailshot but do we need a way to ensure that we don't send these emails twice to the same person? e.g. a field in the database.
- Do we need a rake task to trigger the service class?
- Do we need to parameterise the service class (and rake task) with a datetime for the time that we released the apply again feature or is it ok to assume this will be run soon enough after that happens?

## Link to Trello card

https://trello.com/c/yyvDNpCO/1607-dev-build-email-for-candidates-who-were-unsuccessful-when-we-didnt-have-apply-again

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

malcolmbaig commented on app/services/send_apply_again_call_to_action.rb:
```diff
@@ -0,0 +1,22 @@
+      .joins(:application_choices)
+      .where
+      .not(application_choices: {
+        status: ApplicationStateChange.valid_states - ApplicationStateChange::UNSUCCESSFUL_END_STATES.map(&:to_sym),
```
Can we just query for `where(application_choices: { status: ApplicationStateChange::UNSUCCESSFUL_END_STATES })` ?

stevehook replied:
The intention here is to get applications (forms) for which all the application choices have ended unsuccessfully. So if a candidate got rejected for one choice and accepted an offer for another (or just hasn't heard yet) then we shouldn't be sending this email.

malcolmbaig commented on app/services/send_apply_again_call_to_action.rb:
```diff
@@ -0,0 +1,22 @@
+private
+
+  def unsuccessful_application_forms
+    ApplicationForm
```
I wonder if it's worth explicitly excluding applications that have already started apply_again, in case there's a significant delay between launching the feature and sending the email?

malcolmbaig commented:
> I didn't want to over-engineer this as it's a one-off mailshot but do we need a way to ensure that we don't send these emails twice to the same person? e.g. a field in the database.

Good question - not sure, but we're not doing any handling for this case in other emails, one-shot or otherwise.

>Do we need a rake task to trigger the service class?

I think we're fine to just `rails runner` the service class on the console.

> Do we need to parameterise the service class (and rake task) with a datetime for the time that we released the apply again feature or is it ok to assume this will be run soon enough after that happens?

The latter, IMHO.

stevehook commented:
> > I didn't want to over-engineer this as it's a one-off mailshot but do we need a way to ensure that we don't send these emails twice to the same person? e.g. a field in the database.
> 
> Good question - not sure, but we're not doing any handling for this case in other emails, one-shot or otherwise.
> 
> > Do we need a rake task to trigger the service class?
> 
> I think we're fine to just `rails runner` the service class on the console.
> 
> > Do we need to parameterise the service class (and rake task) with a datetime for the time that we released the apply again feature or is it ok to assume this will be run soon enough after that happens?
> 
> The latter, IMHO.

I've added a condition to the query to exclude applications that have already started an apply again. I also added another condition to avoid the duplicate emails - I had forgotten that we log emails so it's fairly easy to do that.
